common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ErrorHandler.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.shuffle;
+
+import java.net.ConnectException;
+
+/**
+ * Plugs into {@link RetryingBlockFetcher} to further control when an exception should be retried
+ * and logged.
+ * Note: {@link RetryingBlockFetcher} will delegate the exception to this handler only when
+ * - remaining retries < max retries
+ * - exception is an IOException
+ */
+
+public interface ErrorHandler {
+
+  boolean shouldRetryError(Throwable t);
+
+  default boolean shouldLogError(Throwable t) {
+    return true;
+  }
+
+  /**
+   * A no-op error handler instance.
+   */
+  ErrorHandler NOOP_ERROR_HANDLER = t -> true;
+
+  /**
+   * The error handler for pushing shuffle blocks to remote shuffle services.
+   */
+  class BlockPushErrorHandler implements ErrorHandler {
+    /**
+     * String constant used for generating exception messages indicating a block to be merged
+     * arrives too late on the server side, and also for later checking such exceptions on the
+     * client side. When we get a block push failure because of the block arrives too late, we
+     * will not retry pushing the block nor log the exception on the client side.
+     */
+    public static final String TOO_LATE_MESSAGE_SUFFIX =
+        "received after merged shuffle is finalized";
+
+    /**
+     * String constant used for generating exception messages indicating the server couldn't
+     * append a block after all available attempts due to collision with other blocks belonging
```

Comment:
I don't quite get it. Here collision refers to multiple blocks belonging to the same shuffle partition getting pushed at the same time to the same shuffle service. Since the shuffle service needs to completely append one block before handling the other, we get a collision. Right now, all blocks belonging to one shuffle partition get pushed to the same shuffle service, so there won't be collisions between different servers.
Even if we potentially allow multiple shuffle services to handle one shuffle partition in the future, collision is still something that can happen on one server, not between servers.

Did you mean block duplication instead, i.e., the same block getting pushed multiple times? For now, with one shuffle service always handling one shuffle partition, there won't be such cases. If we allow multiple shuffle services to handle one partition down the way, Spark driver should divide blocks based on map Id subranges so that each shuffle service will only handle a disjoint subset of blocks for a given shuffle partition.

Reply:
> Right now, all blocks belonging to one shuffle partition get pushed to the same shuffle service, so there won't be collisions between different servers.

I see. Got it.

Reply:
BTW, how do you know the partition info of a certain shuffle block? Do you use the `mapId` of a shuffle block? If so, I think you should be aware of the new shuffle fetch protocol in 3.0, that we use `taskAttemptId`(which is unique among multiple task attempts) instead of `partitionId` as `mapId` by default. This might break the current assumption of preventing the collision.

Reply:
Yes, we are aware of the distinction between mapId and mapPartitionId introduced in 3.0.
We have internally discussed the implication of this change, and it should be minimum.
SPARK-32923 was created so that we can make push-based shuffle supporting indeterminate stage retries as well.